Initial GKE ingress docs (#584)
akka/akka-platform-guide: docs-source/docs/modules/deployment/pages/gcp-ingress.adoc

> = AWS LoadBalancer Controller

Reviewer: shouldn't this be GKE LoadBalancer Controller ?

Author: Yep, or just "GKE Ingress" as you aren't aware what the controller is called

You need a TLS certificate for the internet facing load balancer. Akka gRPC also needs to be configured with a certificate as GKE ingress for HTTP/2 only works over SSL. This certificate can be any self signed certificate as no validation is done bewtween the loadbalancer the application. // TODO link to and improve the how-to on gRPC TLS

For development and test you can import a self signed certificate, which can be created with for example:

[source,shell script]
----
openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:4096 -keyout dummy-key.pem -out dummy-cert.pem -subj "/CN=dummy/O=dummy"
----

// TODO support uploading to GCP as well
Create a secret for the certificate:

[source, shell script]
----
kubectl create secret tls ingress-secret  --cert dummy-cert.pem  --key dummy-key.pem
----

This is referenced from the Akka Microservice.

== Enable grpcIngress

Add the `grpcIngres` section to the deployment descriptor and `akka.lightbend.com/ingress-class` annotation set to `gce`:

.kubernetes/shopping-cart-service-cr.yml:
[source,yaml]
----
apiVersion: akka.lightbend.com/v1
kind: AkkaMicroservice
metadata:
  name: shopping-cart-service
  namespace: "shopping"
  annotations:

>     akka.lightbend.com/ingress-class: gce

Reviewer 2: Do I understand it correctly that enabling `grpcIngress` without also specifying this annotation (gce or alb) will not work? When we talked about annotations I thought it would be used for custom options, but we could still support the default without them?

Could we add the ingress-class as a property fo the `grpcIngress` section instead so that the default ingress can be enabled at one place?

Reviewer 2: what happens if both grpcIngress and httpIngress are enabled and they need different annotations? I thought the support for annotations would be in grpcIngress/httpIngress sections rather than at the top level?

Author: > what happens if both grpcIngress and httpIngress are enabled and they need different annotations? I thought the support for annotations would be in grpcIngress/httpIngress sections rather than at the top level?

We could have separate annotations ones for http and grpc or put it as a property as you say. Annotations only work in the metadata section. 

If you don't specify any annotations you get a normal Ingress/Service that would work with any Ingress Controller that doesn't require any annotations on the Ingress resource e.g. Minikube Ingress would work. 

Given that you have to install the AWS ingress controller manually that made me think that we shouldn't just default based on the type of cluster we are in.

Reviewer 2: I think we should make the choice (ingress-class) more explicit by making it an explicit property in the grpcIngress/httpIngress sections. We can still have it optional as you say.

I think I misunderstood what we would use annotations for. I thought that was for custom options (AWS LoadBalancer Controller has [massive amount of annotations](https://kubernetes-sigs.github.io/aws-load-balancer-controller/latest/guide/ingress/annotations/) for various things). The operator would just transparently propagate them from the CR to the created Service or Ingress. That is a different story.

Author: > I think we should make the choice (ingress-class) more explicit by making it an explicit property in the grpcIngress/httpIngress sections. We can still have it optional as you say.

Agreed.

> I think I misunderstood what we would use annotations for. I thought that was for custom options (AWS LoadBalancer Controller has massive amount of annotations for various things).

I do use one like that, the gke.akka.lightend.io/backend-config but it is for our behavior, not just forwaded on.

> I thought that was for custom options (AWS LoadBalancer Controller has massive amount of annotations for various things). The operator would just transparently propagate them from the CR to the created Service or Ingress. That is a different story.

I didn't get to that but do agree it will be useful and we need to allow users to specify which annotations go where some how

Reviewer 2: I created a new issue for annotations https://github.com/lightbend/akka-platform-operator/issues/477

= GKE Ingress
:page-toclevels: 3

include::partial$include.adoc[]

> gRPC and HTTP ingress with the https://cloud.google.com/kubernetes-engine/docs/concepts/ingress[GKE Ingress Controller {tab-icon}, window="tab"] is supported
> ion documentation {tab-icon}, window="tab"]

Reviewer 3: ```suggestion
gRPC and HTTP ingress with the https://cloud.google.com/kubernetes-engine/docs/concepts/ingress[GKE Ingress Controller {tab-icon}, window="tab"] is supported  <something seems to be missing here>
<and maybe here too> ion documentation {tab-icon}, window="tab"]
```

== Container native load balancing

Only https://cloud.google.com/kubernetes-engine/docs/how-to/container-native-load-balancing[container native load balancing {tab-icon}, window="tab"] is supported to allow a health check to access the Akka Management port
rather than the gRPC or HTTP ports.

When using HTTP if you override the readiness probe to a path on your HTTP port this will also be used for the load balancer health check.

== TLS certificate

You need a TLS certificate for the internet facing load balancer. Akka gRPC also needs to be configured with a certificate as GKE ingress for HTTP/2 only works over SSL. This certificate can be any self signed certificate as no validation is done bewtween the loadbalancer the application. // TODO link to and improve the how-to on gRPC TLS

For development and test you can import a self signed certificate, which can be created with for example:

[source,shell script]
----
openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:4096 -keyout dummy-key.pem -out dummy-cert.pem -subj "/CN=dummy/O=dummy"
----

// TODO support uploading to GCP as well
Create a secret for the certificate:

[source, shell script]

> ----
> kubectl create secret tls ingress-secret  --cert dummy-cert.pem  --key dummy-key.pem
> ----
> 
> This is referenced from the Akka Microservice.
> 
> == Enable grpcIngress
> 
> Add the `grpcIngres` section to the deployment descriptor with class set to `gce`:
> 
> .kubernetes/shopping-cart-service-cr.yml:
> [source,yaml]
> ----
> apiVersion: akka.lightbend.com/v1
> kind: AkkaMicroservice
> metadata:
>   name: shopping-cart-service
>   namespace: "shopping"
>   annotations:
> spec:
>   image: <image>
>   grpcIngress:
>     enabled: true
>     certificate: "<secret-name>"
>     class: "gce"
> ----

Reviewer 3: ```suggestion
----
kubectl create secret tls my-ingress-secret --cert dummy-cert.pem  --key dummy-key.pem
----

This is referenced from the Akka Microservice.

== Enable grpcIngress

Add the `grpcIngres` section to the deployment descriptor with class set to `gce`:

.kubernetes/shopping-cart-service-cr.yml:
[source,yaml]
----
apiVersion: akka.lightbend.com/v1
kind: AkkaMicroservice
metadata:
  name: shopping-cart-service
  namespace: "shopping"
  annotations:
spec:
  image: <image>
  grpcIngress:
    enabled: true
    certificate: "my-ingress-secret"
    class: "gce"
----
```

When the deployment descriptor has been applied the Akka Operator will create:

> * An Ingress with the appropriate GKE annotations
> * A `ClusterIP` service.
> * A `BackendConfig' to configure a health check to match the readiness check

Reviewer 3: ```suggestion
* an `Ingress` with the appropriate GKE annotations
* a `Service` of type `ClusterIP`
* a https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-features#associating_backendconfig_with_your_ingress[`BackendConfig` {tab-icon}, window="tab"] to configure a health check to match the readiness check
```
